File: src/main/java/com/zepben/vertxutils/routing/handlers/params/BadParamException.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.handlers.params;

import com.zepben.annotations.EverythingIsNonnullByDefault;

@EverythingIsNonnullByDefault
public class BadParamException extends RuntimeException {

    private BadParamException(String msg) {
        super(msg);
    }

    public static BadParamException missingParam(String name) {
        return new BadParamException("missing required parameter " + name);
    }

    public static BadParamException invalidParam(ParamRule<?, ?> rule, String value, String reason) {
        String msg = String.format(
            "Parameter '%s' with value '%s' is invalid. Expected format '%s': %s",
            rule.name(),
            value,
            rule.converter().expectedFormat(),
            reason);

        return new BadParamException(msg);
    }

    public static BadParamException missingBody() {
        return new BadParamException("required body is missing");
    }

    public static BadParamException invalidBody(BodyRule<?> rule, String reason) {
        String msg = String.format(
            "body is invalid. Expected format '%s': %s",
            rule.converter().expectedFormat(),
            reason);

        return new BadParamException(msg);
    }

}

File: src/main/java/com/zepben/vertxutils/routing/handlers/params/BadParamException.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing.handlers.params

class BadParamException private constructor(msg: String) : RuntimeException(msg) {

    companion object {

        fun missingParam(name: String): BadParamException =
            BadParamException("missing required parameter $name")

        fun invalidParam(rule: ParamRule<*, *>, value: String, reason: String?): BadParamException =
            BadParamException("Parameter '${rule.name}' with value '$value' is invalid. Expected format '${rule.converter.expectedFormat}': $reason")

        fun missingBody(): BadParamException =
            BadParamException("required body is missing")

        fun invalidBody(rule: BodyRule<*>, reason: String?): BadParamException =
            BadParamException("body is invalid. Expected format '${rule.converter.expectedFormat}': $reason")

    }

}
